### PR TITLE
Remove baseURI

### PR DESCRIPTION
**Pull request by dcousens**
RE https://github.com/jfurrow/flood/issues/535#issuecomment-344823126

`nginx.conf`
``` nginx
        upstream flood { server 127.0.0.1:3000; }

        # snip

        server {
                listen 80;
                server_name _

                location / {
                        return 200 "<h1>Foobar</h1>";
                        add_header Content-Type text/html;
                }

                location /flood/ { # slash here
                        proxy_pass http://flood/; # and slash here
                        proxy_max_temp_file_size 0; # for downloads > 1.1G
                }
        }
```

Tested and verified locally,  good luck @jfurrow :grinning:

### CODE REVIEW

File: client/src/javascript/stores/ConfigStore.js [hunk 1]
```diff
@@ -26,7 +9,7 @@ class ConfigStoreClass extends BaseStore {
   }
 
   getBaseURI() {
-    return this.userConfig.baseURI || '/';
+    return '';
```
Comment (dcousens) [inline client/src/javascript/stores/ConfigStore.js]: Minimal change was the easiest

**Comment by jfurrow** [comment]
@dcousens I'm super confused... Did you read my comment? baseURI is required for to properly isolate Flood to a subdirectory on a domain.

**Comment by jfurrow** [comment]
Ohhh, my apologies — I see how this could work.

File: client/src/javascript/stores/ConfigStore.js (hunk 1 above)
Comment (jfurrow) [reply client/src/javascript/stores/ConfigStore.js]: Yeah, this PR should be updated to remove all references to `baseURI`.

**Comment by dcousens** [comment]
@jfurrow sorry for lackluster communication on my behalf,  I assumed it might be easier to demonstrate concretely than to make assumptions that I might be wrong about.
This PR simply shows that it works!  But,  open to discussion of course :+1:

**Comment by dcousens** [comment]
> I see how this could work.

tl;dr - URLs are relative if there is no `/` on the front.

**Comment by jfurrow** [comment]
> tl;dr - URLs are relative if there is no / on the front - so it will "just work".

This is only true if the routes aren't nested (they aren't right now, to be fair). As soon as we wish to create a nested route, this will break.

I don't think this is a good idea. It's not difficult for users to specify the `baseURI` option. I'll close this PR, but if you'd like to continue the discussion, you can open a ticket and we can gather more opinions there.

**Comment by dcousens** [comment]
> This is only true if the routes aren't nested (they aren't right now, to be fair). As soon as we wish to create a nested route, this will break.

I don't think I follow,  do you mean something as in `/user/foo/profile`?
The router used on the client won't have any issues handling this as it is handled internally,  not by the server?

**Comment by jfurrow** [comment]
> Any reason not to continue the discussion here?

It seems like this would be more discoverable by users in the future if it were taking place in an issue rather than a PR. But I'm okay with discussing here if that's what you'd prefer.

> The router used on the client won't have any issues handling this as it is handled in the router, not by the server?

The problem is with the requests' relative URLs and nested client-side routes. If the client-side's route is `/user/foo/profile`, and a request is initiated with relative URL `api/auth/logout`, then the actual request URL will be `/user/foo/profile/api/auth/logout`, so it fails. I tested this quickly with your PR by duplicating the `overview` route in `app.js` and changing it to `overview/foo/bar/baz` (of course I also had to modify the generated `index.html` file to use relative URLs). It failed to load, but maybe there's a solution for this — maybe the nginx config needs something else? I'm using the example you provided in this PR.

> I suppose my question is, why do you think the existing approach is a good idea?

So far, the only reason I think this approach is a good idea is because it circumvents the nested client-side route problem.

> I've never had to do this with other services

I'm very much in favor of following any best-practices or commonly implemented patterns for this. I'll do some research, but if you have any examples of documentation around this type of thing, I'd love to read them.

> it is really nice to be able to simply change my nginx configuration and not have to re-compile flood assets, nor need to edit the configuration.

That is understandable, but I honestly don't think it's that bad. Users are required to recompile static assets for every update already.

**Comment by jfurrow** [comment]
One solution to this would be to use [`HashRouter`](https://github.com/ReactTraining/react-router/blob/master/packages/react-router-dom/docs/api/HashRouter.md) instead of [`BrowserRouter`](https://github.com/ReactTraining/react-router/blob/master/packages/react-router-dom/docs/api/BrowserRouter.md), but then we get an ugly `#` in the URL. I'd prefer not to do this.

**Comment by dcousens** [comment]
Thanks for the thorough explanation @jfurrow :+1: ,  I'll check out the issues above and see if there exists a canonical approach :)

**Comment by jfurrow** [comment]
I'm still convinced we need `baseURI`.

**Comment by dcousens** [comment]
I haven't found a better solution yet either :+1:
